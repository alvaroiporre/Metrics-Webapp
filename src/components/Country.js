import React, { useEffect, useRef, useState } from 'react';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import shapes from '../assets/country_shapes.json';
import { getLimits, formData } from '../maps/maps';

const Country = ({ id }) => {
  const indexCountrySelected = useSelector((store) => store.countries.countriesIndexes);
  const index = 0 || indexCountrySelected[id];

  const reference = useRef();

  const [dimensions, setDimensions] = useState({ width: 100, height: 100 });

  const canvasRef = useRef(null);
  // const countryName = id;

  useEffect(() => {
    if (reference.current) {
      setDimensions({
        width: reference.current.offsetWidth,
        height: reference.current.offsetHeight,
      });
    }
    const canvas = canvasRef.current;
    const context = canvas.getContext('2d');
    context.clearRect(0, 0, dimensions.height, dimensions.width);

    let data = shapes[index]?.geo_shape?.geometry?.coordinates;
    if (data === undefined) {
      data = [[[1, 0], [1, 1], [0, 0], [0, 1]]];
    }
    const limits = getLimits(data);
    if (data.length === 1) {
      const polygon = formData(data[0], limits, dimensions);
      context.beginPath();
      context.moveTo(polygon[0].x, polygon[0].y);

      for (let i = 1; i < polygon.length; i += 1) {
        context.lineTo(polygon[i].x, polygon[i].y);
      }

      context.closePath();

      context.fillStyle = 'rgba(255, 0, 0, 0.5)';
      context.strokeStyle = 'black';

      context.fill();
      context.stroke();
    } else {
      data.forEach((poly) => {
        const polygon = formData(poly[0], limits, dimensions);
        context.beginPath();
        context.moveTo(polygon[0].x, polygon[0].y);
        for (let i = 1; i < polygon.length; i += 1) {
          context.lineTo(polygon[i].x, polygon[i].y);
        }

        context.closePath();

        context.fillStyle = 'rgba(255, 0, 0, 0.5)';
        context.strokeStyle = 'black';

        context.fill();
        context.stroke();
      });
    }
  }, [index, dimensions]);

  return (
    <>
      <canvas ref={canvasRef} width={100} height={100} />
    </>
  );
};

Country.propTypes = {
  id: PropTypes.string.isRequired,
};

export default Country;
