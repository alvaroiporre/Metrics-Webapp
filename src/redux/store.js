import { configureStore } from '@reduxjs/toolkit';
import countryReducer from './country/countrySlice';

export default configureStore({
  reducer: {
    countries: countryReducer,
  },
});
