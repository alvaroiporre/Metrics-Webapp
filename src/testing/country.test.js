import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../redux/store';
import Country from '../components/Country';

test('Testing Mission', () => {
  const component = renderer.create(
    <Provider store={store}>
      <Country id="BO" />
    </Provider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
