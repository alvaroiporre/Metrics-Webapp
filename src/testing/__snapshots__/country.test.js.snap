// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Mission 1`] = `
<canvas
  height={100}
  width={100}
/>
`;
